caffe2/perfkernels/sve_emblookup_codegen.py

```diff
@@ -0,0 +1,469 @@
+        code.append(
+            "      for ("
+            + "int64_t"
+            + " j = start_offset; j < end_offset; ++j) {"  # noqa
+        )
+    else:
+        exit(1)
+
+    code.append("        const auto idx = indices[pos];")
+    code.append(
+        "        if (idx < 0 || idx >= data_size) {\n"
+        + "          return false;\n"
+        + "        }"
+    )
+
+    if InType == "uint8_t":
+        code.append("        // unimplemented")
+        code.append("        " + OutType + " wgt = 1.f;")
+        code.append("        // NOLINTNEXTLINE(cppcoreguidelines-init-variables)")
+        code.append("        " + OutType + " bio{};")
+        code.append("        if (weights) {")
+        code.append(
+            "          wgt = weights[IS_WEIGHT_POSITIONAL ? (j - start_offset) : pos];"  # noqa
+        )
+        code.append("        }")
+        if fused:
+            exit(1)
+        else:
+            code.append("        if (scale_bias) {")
+            code.append("          bio = wgt * scale_bias[2 * idx + 1];")
+            code.append("          wgt = wgt * scale_bias[2 * idx];")
+            code.append("        }")
+        code.append("        svfloat32_t vbio = svdup_n_f32(bio);")
+    else:
+        code.append("        " + OutType + " wgt = 1.f;")
+        code.append("        if (weights) {")
+        code.append(
+            "          wgt = weights[IS_WEIGHT_POSITIONAL ? (j - start_offset) : pos];"  # noqa
+        )
+        code.append("        }")
+    code.append("        const svfloat32_t vwgt = svdup_n_f32(wgt);")
+
+    code.append(f"        const {InType}* ip = &input[idx * block_size];")
+
+    # compute and store main loop
+    code.append("        svbool_t pg;")
+    code.append("        for (int64_t k = 0;")
+    code.append("             svptest_first(svAll, pg = svwhilelt_b32_s64("
+                + "k, block_size));")
+    code.append("             k += vLen) {")
+    code.extend(compute(InType, use_weights, isa))
+    code.append("        }\n")
+    code.append("        ++pos;")
+    code.append("      }")
+
+    if use_offsets:
+        code.append("      const int64_t length = end_offset - start_offset;\n")
+        code.append("      if (normalize_by_lengths && length != 0) {")
+        code.append("        const float len_inv = 1.0f / length;")
+    else:
+        exit(1)
+
+    code.append("        svfloat32_t vlen_inv = svdup_n_f32(len_inv);")
+    code.append("        svbool_t pg;")
+    code.append("        for (int64_t j = 0;\n"
+                "             svptest_first(svAll, pg = svwhilelt_b32_s64("
+                "j, block_size));")
+    code.append("             j += vLen) {")
+    code.append(
+        "          svst1_f32(\n"
+        "              pg, &op[j], svmul_f32_x(pg, svld1_f32(pg, &op[j]), vlen_inv));"
+    )
+    code.append("        }")
+    code.append("      }")
+    code.append("    }")
+    return code
+
+
+# start main code
```

Comment:
put the code below under `if __name__ == "__main__":` to avoid running the code when importing this as a library. (This is unlikely to happen but is better practice either way